core/bentley/src/ErrorCategory.ts

```diff
@@ -0,0 +1,117 @@
+/*---------------------------------------------------------------------------------------------
+* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+* See LICENSE.md in the project root for license terms and full copyright notice.
+*--------------------------------------------------------------------------------------------*/
+/** @packageDocumentation
+ * @module Errors
+ */
+
+import { AuthStatus, BentleyError, BentleyStatus, ChangeSetStatus, GeoServiceStatus, HttpStatus, IModelHubStatus, IModelStatus, RepositoryStatus, RpcInterfaceStatus } from "./BentleyError";
+
+/* eslint-disable @typescript-eslint/no-shadow */
+
+/** @alpha */
+export type ErrorCategoryHandler = (error: BentleyError) => ErrorCategory | undefined;
+
+/** A group of related errors for aggregate reporting purposes.
+ * @alpha
+ */
+export abstract class ErrorCategory {
+  public static handlers: Set<ErrorCategoryHandler> = new Set();
+
+  public static for(error: BentleyError): ErrorCategory {
+    for (const handler of this.handlers) {
+      const category = handler(error);
+      if (category) {
+        return category;
+      }
+    }
+
+    return lookupCategory(error);
+  }
+
+  public abstract name: string;
+  public abstract code: number;
+}
+
+namespace HTTP {
+  export enum Successful {
+    OK = 200,
+  }
+
+  export enum ClientError {
+    BadRequest = 400,
+    Unauthorized = 401,
+    Forbidden = 403,
+    NotFound = 404,
+    MethodNotAllowed = 405,
+    NotAcceptable = 406,
+    Conflict = 409,
+    Gone = 410,
+    PreconditionFailed = 412,
+    ExpectationFailed = 417,
+    MisdirectedRequest = 421,
+    UnprocessableEntity = 422,
+    Locked = 423,
+    FailedDependency = 424,
+    UpgradeRequired = 426,
+    PreconditionRequired = 428
+  }
+
+  export enum ServerError {
+    InternalServerError = 500,
+    NotImplemented = 501,
+  }
+}
+
+class Success extends ErrorCategory { public name = "Success"; public code = HTTP.Successful.OK; }
+
+class BadRequest extends ErrorCategory { public name = "BadRequest"; public code = HTTP.ClientError.BadRequest; }
+class Forbidden extends ErrorCategory { public name = "Forbidden"; public code = HTTP.ClientError.Forbidden; }
+class Conflict extends ErrorCategory { public name = "Conflict"; public code = HTTP.ClientError.Conflict; }
+class PreconditionFailed extends ErrorCategory { public name = "PreconditionFailed"; public code = HTTP.ClientError.PreconditionFailed; }
+class UnprocessableEntity extends ErrorCategory { public name = "UnprocessableEntity"; public code = HTTP.ClientError.UnprocessableEntity; }
+
+class Failed extends ErrorCategory { public name = "Failed"; public code = HTTP.ServerError.InternalServerError; }
+class Unknown extends ErrorCategory { public name = "Unknown"; public code = HTTP.ServerError.NotImplemented; }
```

Comment:
From the consumer perspective, NotImplemented would lead me to think that the handler for the request that was made has not been implemented.  Would UnprocessableEntity make sense as a default, considering that if that default is hit, we know that the request did not succeed nor hit any known error categories?

I do understand the perspective that if 'unknown" is hit, something probably wasn't implemented somewhere, but I think it could be confusing to the consumer.

Reply:
I was trying to preserve "UnprocessableEntity"...it has semantics that could apply to more of our iTwin statuses.

The intent with "Unknown" is that it would be almost never used. Only if someone forgets to map a code...
I guess we probably should pick a different HTTP code for that case, though.

Reply:
If we are sure that the error is on our side, I suppose we could go with the infamous 500 (as much as I hate it), especially if we feel that it would be a very rare case.

Reply:
Yes, it would be that rare.

Reply:
I was waiting to see if anyone else chimed in.  Sans other comments and discussion, I would vote for using 500 in this case since it would truly be unexpected.

Reply:
I agree with having a 500 in this instance